src/api/use-subscription.ts

```diff
@@ -25,40 +28,64 @@ export function useSubscription({
   const [state, setState] = useState<SubscriptionState>('NOT_STARTED');
   const [webSocket, setWebSocket] = useState<WebSocket | null>(null);
 
+  const [reconnectCount, setReconnectCount] = useState(0);
+  const [retryCount, setRetryCount] = useState(0);
+  const exponentialBackoff = () => {
+    const delay = Math.pow(2, retryCount) * 1000;
+    return setTimeout(() => setRetryCount(retryCount + 1), delay);
+  };
+
   useEffect(() => {
+    let retryTimeout: NodeJS.Timeout | null = null;
     if (url) {
       const webSocket = new WebSocket(url);
+
       webSocket.onmessage = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         onMessage && onMessage(event);
       };
+
       webSocket.onerror = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         Bugsnag.notify(`WebSocket error "${event.message}" with url: ${url}`);
         onError && onError(event);
       };
+
       webSocket.onclose = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         Bugsnag.leaveBreadcrumb(`WebSocket closed with url: ${url}`);
+
+        // Reconnect immediately if close event is end of stream, otherwise use
+        // exponetial backoff to retry.
+        if (event.code === 1001 && reconnectCount < MAX_NUMBER_OF_RECONNECTS) {
+          setReconnectCount(reconnectCount + 1);
+        } else if (retryCount < MAX_NUMBER_OF_RETRIES) {
```

Comment:
It might be the intention but when the first `if` is evaluated the else is ignored, but it might be the case that both must be evaluated. Is that the case? Or is either reconnect Count and retry?

Reply:
That is as intended. We don't want to retry with exponential backoff if it's a regular reconnect (status code is 1001). So if `status code === 1001` then reconnect directly, otherwise run `exponentialBackoff`.

Maybe not that easy to read, so might rename the function to make it clearer 🤔

Reply:
It might that is not clear the difference between reconnect and retry, but now I see why after explanation.